fhir-cql-server/src/main/java/com/ibm/fhir/cql/engine/server/retrieve/ServerFhirRetrieveProvider.java

```diff
@@ -0,0 +1,148 @@
+
+            Resource resource = executeQuery(dataType, queryParameters);
+            if (resource instanceof Bundle) {
+                AtomicInteger pageNumber = new AtomicInteger(1);
+                FhirBundleCursor cursor = new FhirBundleCursor(url -> {
+                    try {
+                        int nextPage = pageNumber.incrementAndGet();
+                        if( logger.isLoggable(Level.FINE) ) {
+                            logger.fine(String.format("Retrieving page %d / %s", nextPage, url));
+                        }
+                        queryParameters.putSingle(SearchConstants.PAGE, String.valueOf(nextPage));
+                        return resourceHelpers.doSearch(dataType, /*compartment=*/null, /*compartmentId=*/null, queryParameters, DUMMY_REQUEST_URI, /*contextResource=*/null);
+                    } catch (Exception ex) {
+                        throw new RuntimeException(ex);
+                    }
+                }, (Bundle) resource);
+                cursor.forEach(results::add);
+            } else {
+                results.add(resource);
+            }
+        }
+
+        return results;
+    }
+
+    private Resource executeQuery(String dataType, MultivaluedMap<String, String> queryParameters) {
+        Resource resource = null;
+
+        try {
+            if (queryParameters.containsKey("_id")) {
+                String id = queryParameters.getFirst("_id");
+                SingleResourceResult<?> result = resourceHelpers.doRead(dataType, id, true, false, null);
+                if (result.isSuccess()) {
+                    resource = result.getResource();
+                }
+            } else {
+                resource = resourceHelpers.doSearch(dataType, /*compartment=*/null, /*compartmentId=*/null, queryParameters, DUMMY_REQUEST_URI, /*contextResource=*/null);
+            }
+        } catch (RuntimeException rex) {
+            throw rex;
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+
+        return resource;
+    }
+
+    private MultivaluedMap<String, String> getQueryParameters(SearchParameterMap map) {
+
+        MultivaluedMap<String, String> parameters = new MultivaluedHashMap<>();
+        for (Map.Entry<String, List<List<? extends IQueryParameter>>> entry : map.entrySet()) {
+            for (List<? extends IQueryParameter> sublist : entry.getValue()) {
+                for (IQueryParameter p : sublist) {
+                    String name = getModifiedName(entry.getKey(), p);
+
+                    List<String> values = parameters.computeIfAbsent(name, k -> new ArrayList<>());
+                    values.add(p.getParameterValue());
+                }
+            }
+        }
+        if( map.count() != null ) {
+            parameters.putSingle(SearchConstants.COUNT, String.valueOf(map.count()));
+        }
+        return parameters;
+    }
+
+    protected String getModifiedName(String name, IQueryParameter param) {
```

Comment:
I'm almost positive we have some code for this in SearchUtil.  Maybe expose that via public static method?

Reply:
This version is specific to the QueryParameter objects that I created for the SearchParameterResolver class in the fhir-cql module. It needs to stay separate.

Reply:
how is this different from fhir-cql-rest/src/main/java/com/ibm/fhir/cql/engine/rest/retrieve/RestFhirRetrieveProvider.java

Reply:
If you are asking about the particular method getModifiedName(), that looks like a copy/paste job. I can move it up into the SearchParameterFhirRetrieveProvider base class.